### ISSUE TITLE
`get_area_slices` method returns wrong slices

### ISSUE THREAD

**Issue by BENR0** [pytroll/pyresample#371]
If a geos area is sliced with the results of `get_area_slices` called with an area which is a subset of the area the resulting area is bigger than the subset area.

```python
from satpy.resample import get_area_def

a = get_area_def("seviri_0deg")
b = a[slice(54, 822), slice(1461, 2485)]
print(b)
c = a.crop_around(b))
print(c)
a.get_area_slices(b)
```
I would expect b and c to have the same number of rows/columns.
I think the problem is the rounding in `get_area_slices`:

https://github.com/pytroll/pyresample/blob/2456508a6d18e04dcce517a0c09238fd5e7bb0da/pyresample/geometry.py#L2201

This in turn affects the scene `crop` method in Satpy. Interestingly while the scene `resample` method also uses `get_area_slices` for data reduction at least the nearest neighbour resampler "corrects" the wrong slices somewhere.

**Comment by djhoese** [comment]
Just for reference, what do the prints in your example code show?

**Comment by BENR0** [comment]
@djhoese the just print the area so one can see the width/height and extent for comparison. you can go without if you want and just compare the slices.

**Comment by djhoese** [comment]
Yeah, I just wanted to document what the actual sizes are that are being produced:

First: 

```
Area ID: seviri_0deg
Description: Full disk MSG image 0 degrees
Projection: {'a': '6378169', 'h': '35785831', 'lon_0': '0', 'no_defs': 'None', 'proj': 'geos', 'rf': '295.488065897001', 'type': 'crs', 'units': 'm', 'x_0': '0', 'y_0': '0'}
Number of columns: 1024
Number of rows: 768
Area extent: (-1186659.4967, 3103917.1917, 1885753.4606, 5408226.9096)
```

Second:

```
Area ID: seviri_0deg
Description: Full disk MSG image 0 degrees
Projection: {'a': '6378169', 'h': '35785831', 'lon_0': '0', 'no_defs': 'None', 'proj': 'geos', 'rf': '295.488065897001', 'type': 'crs', 'units': 'm', 'x_0': '0', 'y_0': '0'}
Number of columns: 1025
Number of rows: 770
Area extent: (-1189659.9, 3100916.7884, 1885753.4606, 5411227.3129)
```

Slice:

```
(slice(1460, 2485, None), slice(53, 823, None))
```

**Comment by djhoese** [comment]
:thinking: Hm looking at the code this may be a logic bug. The `_get_slice_starts_stops` method gets the floating point array indices of the extents of the area to cover. We want our result to completely include that area. But our indexes, I think, should be shifted half a pixel because the indexes used for the AreaDefinition are the centers of the pixels. So it should probably be `round(x[0] + 0.5)` for example.

**Comment by BENR0** [comment]
Yes, I think that is the logic why on the stop side a 1 is added already. So I guess something like this would work for the start side. I am not sure if this is a solution for all cases. The problem I think is that `round` will floor for values <= x.5 which is the case in this scenario if I remember right (tested this a while back and only have come around to make an issue now).

Not sure if this produces the correct results in all cases but I was thinking maybe not to round but to use ceil? Another possibility I was thinking about to just round without adding anything to the array indices and make sure that `get_array_coordinates_from_projection_coordinates` always returns indices > x.5 by adding/subtracting for example area.pixel_size_x/2 / area.pixel_size_y/2 to the area extent? Maybe in the end these are all equal to what you propose and have no advantage. So just adding 0.5 to the indices before rounding would be the easiest.

**Comment by djhoese** [comment]
> I think that is the logic why on the stop side a 1 is added already

That logic is because python's indexing is exclusive on the stop index. Example: `[1, 2, 3, 4, 5][:3] -> [1, 2, 3]`

> but I was thinking maybe not to round but to use ceil?

I thought about that for a second and then realized that using floor or ceil explicitly will likely always produce less or more pixels than intended (further away from the area of coverage).

> and make sure that get_array_coordinates_from_projection_coordinates always returns indices > x.5 by adding/subtracting for example area.pixel_size_x/2 / area.pixel_size_y/2 to the area extent?

The `get_array_coordinates` function needs to always return the straight conversion from projection coordinates to array coordinates. They are floating point numbers because of this. No `0.5` add/subtract magic should be added in that method.

**Comment by djhoese** [comment]
@mraspaud was the last person to work on this stuff and may have ideas, but he's on vacation right now.

**Comment by BENR0** [comment]
> The `get_array_coordinates` function needs to always return the straight conversion from projection coordinates to array coordinates. They are floating point numbers because of this. No `0.5` add/subtract magic should be added in that method.

Half the pixel size could just be added before calling `get_array_coordinates` so no need to change the method itself. Never the less the easiest/clearest options should be used.

> @mraspaud was the last person to work on this stuff and may have ideas, but he's on vacation right now.

Ok. There is no hurry from my side. We currently just account for the pixel shift.
